packages/ra-core/src/form/useInitializeFormWithRecord.ts

```diff
@@ -19,10 +19,13 @@ const useInitializeFormWithRecord = record => {
         // Disable this option when re-initializing the form because in this case, it should reset the dirty state of all fields
         // We do need to keep this option for dynamically added inputs though which is why it is kept at the form level
         form.setConfig('keepDirtyOnReinitialize', false);
-        // Ignored until next version of final-form is released. See https://github.com/final-form/final-form/pull/376
-        // @ts-ignore
-        form.restart(initialValuesMergedWithRecord);
-        form.setConfig('keepDirtyOnReinitialize', true);
+        // Since the submit function returns a promise, use setTimeout to prevent the error "Cannot reset() in onSubmit()" in final-form
+        setTimeout(() => {
```

Comment:
This is really concerning to me. Any error in the form.restart() won't be caught. 

Can you explain why the error occurs, why the setTimeout removes it, and check if there is no other way? And can you add a unit test?

Reply:
This is the way `final-form` is telling to do it: https://github.com/final-form/final-form/blob/ee1ef7272882a966644ca1bf0ea6f115a2492251/src/FinalForm.js#L933.
However this constraint seems to have been removed in the next version:
https://github.com/final-form/final-form/pull/363
I don't see how we could add a unit test for this.

Reply:
I've added a comment to explain that the `setTimeout` should be removed when the next version of `final-form` will be released.
And actually there is already a test for this:
https://github.com/marmelab/react-admin/blob/baaf3da901ccd68dc8d42f0a2747a9f78e8eadff/cypress/integration/edit.js#L260-L277
(it throws the linked error without the `setTimeout`)